Add postNL letters

scheric opened the pull request:
## Description:
This change will add a new postNL letter entity to fronted. it is using the same library so i added it to the existing file.

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: postnl
    username: POSTNL_USERNAME
    password: POSTNL_PASSWORD
    letter: (true / falce)
```

## Testing:
  - [x] The code change is tested and works locally.

I can not test what happens if you did not enable the postNL letter service en do enable the entity.

## Documentation 
If user exposed functionality or configuration variables are added/changed:
home-assistant/home-assistant.github.io#5731 
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

- [ ]  waiting for approval / review


backup before big file change 
[Nieuw tekstdocument.txt](https://github.com/home-assistant/home-assistant/files/2181194/Nieuw.tekstdocument.txt)


[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

scheric committed:
created a new letters entity

scheric committed:
making entity go to frontend

Creating an if statement so you can easily enable and disable the letters entity in configuration.yml.

scheric committed:
adding an icon

Adding an icon for letters and changing the name of the icon in the package entity.

scheric committed:
standard unit mesurement

homeassistant commented:
Hi @scheric, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#15389).

Once you do that we will be able to review and accept this pull request.

Thanks!

scheric committed:
resolved some houndci-bot errors

scheric commented on homeassistant/components/sensor/postnl.py:
```diff
@@ -50,7 +55,10 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
         return
 
-    add_devices([PostNLSensor(api, name)], True)
+    if letter == True:
```
I am not so experienced to understand what you are saying.
should i change it to:
```
    if letter:
        add_devices([PostNLSensor(api, name), PostNLletter(api, name)], True)
    else:
        add_devices([PostNLSensor(api, name)], True)
```

source: https://lintlyci.github.io/Flake8Rules/rules/E712.html

scheric committed:
made valid names for constants

And an additional fix for houndci-bot.

hmmbob commented:
Cool, looking forward to testing this. 

Please note a typo in the code: "ICON_PACKEGE" --> "ICON_PACKAGE" on line 27 and line 97.

scheric committed:
spelling mistake

scheric committed:
more spelling mistakes

scheric commented on homeassistant/components/sensor/postnl.py:
```diff
@@ -107,3 +115,50 @@ def update(self):
         self._state = len(status_counts)
+
+
+class PostNLletter(Entity):
```
Is there any information on https://developers.home-assistant.io ore any other site that i can follow because i cant get it working. 

I am getting this error in HA. It is a useless error for me so i am kind off stranded here.
```
SyntaxError: invalid syntax
Fatal error while loading config: invalid syntax (postnl.py, line 73)
Failed config
  General Errors: 
    - invalid syntax (postnl.py, line 73)
Successful config (partial)
```

This is what i tried to do. 
[postnl.txt](https://github.com/home-assistant/home-assistant/files/2179977/postnl.txt)
[postnl.zip](https://github.com/home-assistant/home-assistant/files/2179996/postnl.zip) (includes postnl.py)

fabaff replied:
Your configuration is not passing the validation.

https://developers.home-assistant.io/docs/en/development_validation.html

scheric replied:
Thanks. I really appreciate your help. 👏

bramkragten commented:
Can we also add a postnl camera? Like the [USPS Camera](https://www.home-assistant.io/components/camera.usps/)?

scheric committed:
making from two classes one class

things that are commented in this version are not in use

scheric committed:
solving houndci-bot errors

scheric committed:
solving more houndci-bot errors

scheric committed:
making a visual indentation for houndci-bot

scheric committed:
solving more houndci-bot errors

scheric commented:
I have two different entity's in HA. with different names, units and icons but how can i give them different states.  
I did not find any info in HA dev docs

fabaff commented:
Take a look at the [`systemmonitor` sensor](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/systemmonitor.py#L113) for one way to do it.

scheric committed:
states now react to different entity

scheric commented:
Is there a need to have the possibility to give the sensors different names in configuration.yaml

possible example
```yaml
sensor:
  - platform: postnl
    packagename: USER_SET_NAME1
    lettername: USER_SET_NAME2
    username: POSTNL_USERNAME
    password: POSTNL_PASSWORD
    letter: (true / falce)
```

fabaff commented on homeassistant/components/sensor/postnl.py:
```diff
@@ -24,14 +24,15 @@
-
 MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=30)
 
+CONF_LETTER = 'letter'
```
Because it would require a definition in `const.py`. If only one platform is using it then it doesn't make much sense to add a constant to that file..

fabaff commented:
> Is there a need to have the possibility to give the sensors different names in configuration.yaml

Usually we allow to set a `name:` in the configuration to over-write the default name and add the sensor type to that.

fabaff commented on homeassistant/components/sensor/postnl.py:
```diff
@@ -50,18 +52,36 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
 
-    add_devices([PostNLSensor(api, name)], True)
+    sensor_items = []
+    sensor_items.extend([PostNLSensor(
```
Use `.append(...)` instead of `.extend([...])`.

fabaff commented on homeassistant/components/sensor/postnl.py:
```diff
@@ -50,18 +52,36 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
-    add_devices([PostNLSensor(api, name)], True)
+    sensor_items = []
+    sensor_items.extend([PostNLSensor(
+        hass,
```
No need to pass `hass`. You are not using it later.

fabaff commented on homeassistant/components/sensor/postnl.py:
```diff
@@ -50,18 +52,36 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
+    sensor_items = []
+    sensor_items.extend([PostNLSensor(
+        hass,
+        'PostNL-package',
```
`PostNL` is the part of the default name. The sensor is called `package` or `packages`. Create the name later while you initialize the sensor with something like `self._name = '{} {}'.format(name, sensor_type)` while `name` is the default name or the one that was defined by the user.

fabaff commented on homeassistant/components/sensor/postnl.py:
```diff
@@ -50,18 +52,36 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
+        hass,
+        'PostNL-package',
+        'package-variant-closed',
+        'packeges',
```
Typo

fabaff commented on homeassistant/components/sensor/postnl.py:
```diff
@@ -50,18 +52,36 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
         self._attributes = None
         self._state = None
         self._api = api
+        self._icon = "mdi:" + icon
```
Please use string formatting -> `'mdi:{}'.format(icon)`

Or simpler: Just pass the icon name to the constructor, e.g., `mdi:email-outline` and `mdi:package-variant-closed`.

scheric committed:
minor changes and a bit off cleanup

added a configurable name, repaired a spelling mistake, removed unused stuff, added sensor type.

scheric commented:
@iMicknl, that is indeed a good plan.
I am not able to import any state that `validate_letters()` gives.
I tried this: `self._api.validate_letters()` but i am getting a `AttributeError: 'NoneType' object has no attribute` error. 

## In postnl_api.py

`raise Exception(validation)` results into `Exception: {'status': 'Validated'}`
So i need to use a tuple.
but if i do this in postnl.py
```python
test = self._api.validate_letters()
_LOGGER.warning(test[1])
```
I am still getting an `AttributeError: 'NoneType' object has no attribute` error

What am i doing wrong here?

iMicknl commented:
@scheric that's likely a bug in the api wrapper package itself. Those letter functions haven't been tested thoroughly yet, since they aren't used anywhere.

scheric commented:
If i use this in the setup_platform.
 ```python
validation = self._api.validate_letters()
_LOGGER.warning(validation["status"]


```
It will result to  `AttributeError: 'NoneType' object has no attribute 'validate_letters'`
but if i use the same code in `class PostNLSensor` in the `update` function i receive `Validated` in the console. Why can i only use this in the `class PostNLSensor` and not in the setup?

I made some little changes in the postnl.py so here is it if you want to have a look.
[postnl3.txt](https://github.com/home-assistant/home-assistant/files/2213999/postnl3.txt)

balloob commented:
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it.
